sources/src/main/java/com/google/solutions/jitaccess/core/MailAddressRegexFormatter.java

```diff
@@ -0,0 +1,58 @@
+package com.google.solutions.jitaccess.core;
+
+import java.util.regex.Pattern;
+import java.util.stream.IntStream;
+
+import com.google.common.base.Preconditions;
+
+public class MailAddressRegexFormatter implements MailAddressFormatter {
+    private final Options options;
+
+    public MailAddressRegexFormatter(Options options) {
+        this.options = options;
+    }
+
+    public record Options(Pattern internalsMailAddressPattern, String internalsMailAddressTransform,
+            Pattern externalsMailAddressPattern, String externalsMailAddressTransform) {
+        public Options {
+            Preconditions.checkNotNull(internalsMailAddressPattern);
+            Preconditions.checkNotNull(internalsMailAddressTransform);
+        }
+    }
+
+    public Options options() {
+        return this.options;
+    }
+
+    @Override
+    public String format(String mailAddress) {
+        var formattedAddress = transform(mailAddress, options().internalsMailAddressPattern,
+                options().internalsMailAddressTransform);
+        if (formattedAddress != null) {
+            return formattedAddress;
+        }
+
+        if (options.externalsMailAddressPattern == null) {
+            throw new IllegalArgumentException("Email address does not conform to configured patterns.");
+        }
```

Comment:
Would it make sense to make the error message here different for the two `throws` statement, just for easier debugging purposes?

Reply:
Sure, makes sense.

Reply:
Changed.